### ISSUE TITLE
feature Request -- MQTT support

### ISSUE THREAD

**Issue by bartekd123** [gentlepumpkin/bi-aidetection#11]
Would be great to have MQTT support so that it sends the alert along with what was detected. Would that be possible? That way the data can be parsed, and create customized notifications (EG using NodeRed and home-assistant)

**Comment by msauto** [comment]
+1 on MQTT ! Please! It would also be great if the log could be monitored some how and send an alert if errors present or program becomes unresponsive.

**Comment by AndrewHoover** [comment]
This would be very helpful especially being able to reference the picture name that triggered the event so that another service can do something with it like include the picture as part of a notification.

**Comment by gentlepumpkin** [comment]
the following placeholders were added in code recently:
[detection]
[position]
[camera]
[confidence]
[detections] (all detected objects separated by comma)
[confidences] (all confidences separated by comma in the same oder as above)

**Comment by AndrewHoover** [comment]
Would any of these yield the image name that was used to produce the positive detection?

**Comment by gentlepumpkin** [comment]
no, but that could be added. What do you need it for?

**Comment by AndrewHoover** [comment]
I need a way to pull the actual file name of the image into my home automation system so that it can then fire events with the image as a payload such as notifications and archiving.

**Comment by gentlepumpkin** [comment]
just wrote an update. [imagepath] will give you the full path to the image that caused the alert. Would you be so kind to build the latest version an check if it works?

**Comment by AndrewHoover** [comment]
Sure. It may take some time for me to figure out how to get that done...

**Comment by AndrewHoover** [comment]
I'm not a C# coder so I'm not setup to compile this but I'm working on it.
In the meantime, quick question, does the [imagepath] include the absolute path as referenced by AITool or just the file name? It may be more beneficial to also offer just the file name as the path on the remote host will not see the same path?

Thank you!

**Comment by gentlepumpkin** [comment]
Voila:
[aitool.zip](https://github.com/gentlepumpkin/bi-aidetection/files/4753671/aitool.zip)

**Comment by gentlepumpkin** [comment]
Now the following option are available:
[imagepath] gives the full path
[imagefilename] gives only the filename

Please test them out well :D

**Comment by AndrewHoover** [comment]
Excellent! Thank you!
Ok, question on usage, I believe I saw that the Trigger URL(s) are separated by a comma(?). If so, in the url for this usage, does this example properly demonstrate the implementation of the placeholders listed above?

http://url.myhome.server/handler?image=[imagefilename]&detection=[detection]&position=[position]

Thanks again!

**Comment by gentlepumpkin** [comment]
This URL should work. 
Yes, the Trigger URLs are separated by a comma. When entering multiple trigger urls you can also hit space after the first URL and the comma will be auto-created.

**Comment by gentlepumpkin** [comment]
Nice website btw

**Comment by AndrewHoover** [comment]
Awesome! I'll write up a little script this evening to parse the parameters. So far, I'm not seeing any change in normal behavior. 

Thanks on the site complement.

**Comment by bartekd123** [comment]
Thanks @gentlepumpkin i can test it out tonight too. Where do you configure the MQTT url as well as what the topic that it gets sent to? Or did you just add it in code for the future?

**Comment by gentlepumpkin** [comment]
all the placeholders that were mentioned in this issue can be added to a normal trigger URL. They are added to the code and you can download the latest build of aitool a few posts above (aitool.zip)

**Comment by bartekd123** [comment]
Thanks, I did download the newest aitool. When I open it, there are no options for MQTT. Is there some way I can send a MQTT message using the trigger url? Or you added the ability to add the placeholders to a standard http url?

**Comment by gentlepumpkin** [comment]
this is a standard http url. I have no experience with MQTT and thus don't really know how it works. But isn't it possible to contact home assistant with a http url?

**Comment by AndrewHoover** [comment]
@bartekd123 if you have some experience coding, you might be able to setup a PHP script to accept the variables via url and then post them via MQTT. I am working on a simple script to test these variables based on this library:
https://github.com/php-mqtt/client

**Comment by AndrewHoover** [comment]
Another example for using PHP:
https://www.hivemq.com/blog/mqtt-client-library-encyclopedia-mosquitto-php/

**Comment by AndrewHoover** [comment]
@gentlepumpkin Ok, I wrote up a simple php script to capture the query sent from AITool using the following URL pattern:
test.php?image=[imagefilename]&detection=[detection]&position=[position]
and it looks like the imagefilename variable is still getting the full path in addition to the file name:

image=B:%5CBlueIris%5Caiinput%5CBackPorchSD.20200609_182244337.jpg&detection=person&position=21,319,90,480
image=B:%5CBlueIris%5Caiinput%5CBackPorchSD.20200609_182249392.jpg&detection=person&position=211,292,253,368
image=B:%5CBlueIris%5Caiinput%5CBackPorchSD.20200609_182416778.jpg&detection=person&position=158,307,206,435

**Comment by AndrewHoover** [comment]
Additionally when I changed the [imagefilename] over to [imagepath] I did get the full path:
image=B:%5CBlueIris%5Caiinput%5CBackPorchSD.20200609_183218923.jpg&detection=person&position=183,301,220,404
image=B:%5CBlueIris%5Caiinput%5CBackPorchSD.20200609_183223928.jpg&detection=person&position=210,293,243,366

**Comment by bartekd123** [comment]
thanks @AndrewHoover  Ill play around with what you mentioned.

Thanks @gentlepumpkin  for your hard work on this app

**Comment by gentlepumpkin** [comment]
I fixed the issue getting the full path instead of the file name. I published 1.67 pre, it should work now.
[Download v1.67-pre](https://github.com/gentlepumpkin/bi-aidetection/releases/download/v1.67-pre/AI.Tool.1.67-pre.zip)

**Comment by europer** [comment]
yes, a great program and easy to use and works perfectly, just one thing as stated above would be great if just to have one payload and topic (to define) as mqtt. The problem ist that blue iris do not recon the trigger of you programm (even if its trigger) and do not in alert send a mqtt to my server (in my case a node-red) just when I use motion zones, but that is more or less the core issue not to use this but your AI. the nice thing is it stays at home all information and to take the info back from telegram into node-red (or what ever) is like a looooooong way from my internal server....long speach short meaning...would it be possible to implemtent one payload and topic to a mqtt broker?
thanks in advance

**Comment by ipeterski** [comment]
hi, so i was watching the mqtt for the trigger and cancels.. my cameras are not getting a cancel request.. so i watched as a car drove , ai tools send a trigger command, yet it never sent the cancel,,, the cancel fields are red text, and not black, not sure if that means anything.. i will have a look in the config to see anything, but not sure how the cancel fields are activated... and again im refering to the mqtt cancel fields.. the url trigger and cancel work fine. but i would like to utilize the mqtt.. 

thanks in advanced.
 --Stickers
